---
title: "Sealion single locus metylation report "
output: html_notebook
author: "charles-gumbi)"
date: "2025-05-19"
---

### Loading necessary libraries
```{r libraries}
library(dplyr)
library(stringr)
library(paletteer)
library(ggplot2)
library(tidyverse)
library(RcppParallel)
library(tidyr)
library(multidplyr)
library(emmeans)
library(rstanarm)
library(coda)
library(igraph)

#install.packages("igraph", dependencies = TRUE)
#install.packages("rstanarm", dependencies = TRUE)
```

### Defining the directory containing Bismark reports
```{r directiories}
report_dir <- "~/methylation/sea_lion/bismark_output/locus001/alignment/"
report_files <- list.files(report_dir, pattern = "_bismark_bt2_PE_report.txt$", full.names = TRUE)
```

### Fuction to extract the relevent infomation from ech individual file 
```{r fuctions}
# Initialize an empty list to store extracted data
report_list <- list()

# Loop through all report files and extract relevant values
for (file in report_files) {
  # Read the file as plain text
  file_content <- readLines(file)
  
  # Extract relevant alignment statistics
  TotalPairs <- as.numeric(str_extract(file_content[grep("Sequence pairs analysed in total:", file_content)], "\\d+"))
  UniqueAlignments <- as.numeric(str_extract(file_content[grep("Number of paired-end alignments with a unique best hit:", file_content)], "\\d+"))
  MappingEfficiency <- as.numeric(str_extract(file_content[grep("Mapping efficiency:", file_content)], "\\d+\\.\\d+"))
  NoAlignment <- as.numeric(str_extract(file_content[grep("Sequence pairs with no alignments under any condition:", file_content)], "\\d+"))
  NotUnique <- as.numeric(str_extract(file_content[grep("Sequence pairs did not map uniquely:", file_content)], "\\d+"))
  Discarded <- as.numeric(str_extract(file_content[grep("Sequence pairs which were discarded because genomic sequence could not be extracted:", file_content)], "\\d+"))
  TotalNumberOfCsAnalysed <- as.numeric(str_extract(file_content[grep("Total number of C's analysed", file_content)], "\\d+"))
  TotalMethylatedCsInCpGCotext <- as.numeric(str_extract(file_content[grep("Total methylated C's in CpG context:", file_content)], "\\d+"))
  TotalMethylatedCsInCHGCotext <- as.numeric(str_extract(file_content[grep("Total methylated C's in CHG context:", file_content)], "\\d+"))
  TotalMethylatedCsInCHHCotext <- as.numeric(str_extract(file_content[grep("Total methylated C's in CHH context:", file_content)], "\\d+"))
  TotalMethylatedCsInUnknownCotext <- as.numeric(str_extract(file_content[grep("Total methylated C's in Unknown context", file_content)], "\\d+"))
  TotalUnmethylatedCsInCpGCotext <- as.numeric(str_extract(file_content[grep("Total unmethylated C's in CpG context", file_content)], "\\d+"))
  TotalUnmethylatedCsInCHGCotext <- as.numeric(str_extract(file_content[grep("Total unmethylated C's in CHG context", file_content)], "\\d+"))
  TotalUnmethylatedCsInCHHCotext <- as.numeric(str_extract(file_content[grep("Total unmethylated C's in CHH context", file_content)], "\\d+"))
  TotalUnmethylatedCsInUnknownCotext <- as.numeric(str_extract(file_content[grep("Total unmethylated C's in Unknown context", file_content)], "\\d+"))
  CmethylatedInCpGcontext <- as.numeric(str_extract(file_content[grep("C methylated in CpG context", file_content)], "\\d+"))
  CmethylatedInCHGcontext <- as.numeric(str_extract(file_content[grep("C methylated in CHG context", file_content)], "\\d+"))
  CmethylatedInCHHcontext <- as.numeric(str_extract(file_content[grep("C methylated in CHH context", file_content)], "\\d+"))
  
  
  # Get sample name
  sample_name <- gsub("_bismark_bt2_PE_report.txt", "", basename(file))
  
  # Store extracted data in a list
  report_list[[sample_name]] <- data.frame(
    Sample = sample_name, 
    TotalPairs = TotalPairs, 
    UniqueAlignments = UniqueAlignments, 
    MappingEfficiency = MappingEfficiency, 
    NoAlignment = NoAlignment, 
    NotUnique = Discarded, 
    Discarded = Discarded,
    TotalNumberOfCsAnalysed=TotalNumberOfCsAnalysed,
    TotalMethylatedCsInCpGCotext=TotalMethylatedCsInCpGCotext,
    TotalMethylatedCsInCHGCotext=TotalMethylatedCsInCHGCotext,
    TotalMethylatedCsInCHHCotext=TotalMethylatedCsInCHHCotext,
    TotalMethylatedCsInUnknownCotext=TotalMethylatedCsInUnknownCotext,
    TotalUnmethylatedCsInCpGCotext=TotalUnmethylatedCsInCpGCotext,
    TotalUnmethylatedCsInCHGCotext=TotalUnmethylatedCsInCHGCotext,
    TotalUnmethylatedCsInCHHCotext=TotalUnmethylatedCsInCHHCotext,
    TotalUnmethylatedCsInUnknownCotext=TotalUnmethylatedCsInUnknownCotext,
    CmethylatedInCpGcontext=CmethylatedInCpGcontext,
    CmethylatedInCHGcontext=CmethylatedInCHGcontext,
    CmethylatedInCHHcontext=CmethylatedInCHHcontext)
}
# Combine all extracted data into a single dataframe
final_dataframe <- bind_rows(report_list)

# Save the combined dataframe as a CSV file
write.csv(final_dataframe, "~/methylation/sealion_clock/output_csv/combined_bismark_alignment_summary.csv", row.names = FALSE)
```

### Creating summary statistics
```{r summmary}

summary_stats <- final_dataframe %>%
  summarise(
    Sample  = n(),
    Mean_Mapping_Efficiency = mean(MappingEfficiency , na.rm = TRUE),
    SD_Mapping_Efficiency = sd(MappingEfficiency , na.rm = TRUE),
    Mean_CpG_Methylation = mean(CmethylatedInCpGcontext, na.rm = TRUE),
    SD_CpG_Methylation = sd(CmethylatedInCpGcontext, na.rm = TRUE),
    Mean_Total_Reads = mean(TotalPairs , na.rm = TRUE),
    SD_Total_Reads = sd(TotalPairs , na.rm = TRUE))
# Print summary statistics
print(summary_stats)
```



### Mapping efficiency plot
```{r mapping_plot}
mapping_plot <- ggplot(final_dataframe, aes(x = Sample, y = MappingEfficiency)) +
  geom_bar(stat = "identity") +
  theme_minimal() +
  labs(title ="Mapping Efficiency Across Samples",
       y = "Mapping Efficiency (%)",
       x = "Sample") + 
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(colour="black", size=12),
        panel.border = element_rect(colour="black", fill = NA, size=1), 
        panel.background = element_blank(), 
        plot.title = element_text(hjust=0.5, size=15),
        text = element_text(size = 12,  family = "sans"))
  
mapping_plot
ggsave(plot = mapping_plot, filename = "~/methylation/sealion_clock/output_plot/mapping_plot.png" , width = 8, height = 6, dpi = 300)
```

###  Visualizing methylation levels in three context (i.e. CpG, CHH, CHG) across samples. 
```{r meltylation_level_plot}
methylation_long <- final_dataframe %>%
    pivot_longer(cols = c(CmethylatedInCHHcontext, CmethylatedInCHGcontext, CmethylatedInCpGcontext),
                 names_to = "Cytosine_context",
                 values_to = "Methylation")

methylation_plot <- ggplot(methylation_long, aes(x = Sample, y = Methylation, fill = Cytosine_context)) +
  geom_bar(stat = "identity", position = "dodge") +
  theme_minimal() +
  labs(title = "Methylation level", 
       y = "Methylation (%)",
       x = "Sample", 
       fill = "Cytosine context") +
  theme(axis.text.x = element_blank(),
        axis.text.y = element_text(colour="black", size=12),
        panel.border = element_rect(colour="black", fill = NA, size=1), 
        panel.background = element_blank(), 
        plot.title = element_text(hjust=0.5, size=15),
        text = element_text(size = 12,  family = "sans"))
methylation_plot

ggsave(plot = methylation_plot, filename = "~/methylation/sealion_clock/output_plot/methylation_plot.png", width = 8, height = 6, dpi = 300)
```

### Load the matrix of percent methylation values

```{r adding_metadata}
metadata <- read.csv("~/methylation/sealion_clock/output_csv/Ejubatus_ABLG_export_20250506.csv", header = T)
# Remove "_L001_R1_trimmed" from sample names

sealion_df <- final_dataframe %>%
  mutate(Sample = gsub("_L001_R1_trimmed", "", Sample)) %>%
   select(Sample, CmethylatedInCpGcontext) %>%
  separate(Sample, into = c("ABLG", "PlateNumber"), sep = "_") %>%
  mutate(loci001 = CmethylatedInCpGcontext / 100)

sealion_df <- merge(sealion_df, metadata, by = "ABLG") %>%
  select(ABLG, loci001, AgeYRS)
```

### Calculate Pearson's correlations

```{r pearson_corr}
corr_matrix <- corr.test(x = as.matrix(sealion_df[,-2]), 
                         y = sealion_df[,2], use="pairwise",
                         method="pearson", adjust="BH", alpha=0.05)

#Format dataframe

corr_results <- data.frame(corr_matrix$r, corr_matrix$p.adj)
corr_results <- tibble::rownames_to_column(corr_results, "Loc")
colnames(corr_results) <- c("Loc", "r", "padj")
corr_results$absr <- abs(corr_results$r)

#Extract CpG sites with top 10% of Pearson's correlations

quantile(corr_results$absr, 0.90)

loc_highr <- corr_results %>%
  filter(absr > 0.4465467)

BBRF_percent_wide_filt <- BBRF_percent_wide %>%
  select(loc_highr$Loc, Age)

#Run cv.glmnet

CVGLM <- cv.glmnet(x = as.matrix(BBRF_percent_wide_filt[,-1015]),
                   y = BBRF_percent_wide_filt[,1015],
                   nfolds = nrow(BBRF_percent_wide_filt),
                   alpha = 0,
                   type.measure = "mae",
                   family = "gaussian",
                   grouped=FALSE)

min(CVGLM$cvm)




```




